### PR TITLE
Cap how many times an answer modifier charges per order

### PR DESCRIPTION
**Pull request by stefan-burke**
## What changed

A question-answer modifier charged once per ticket that picked its
answer, so a delivery-area fee multiplied across a multi-ticket order
(3 tickets x £10 Zone 2 = £30). Owners can now set **Max times per
order** on an answer-triggered modifier: 1 charges a delivery fee once
per booking, N caps the charge at N, blank keeps per-ticket charging.

Stacked on #2150 (the guide recipe this completes); now targeting main
after that merged.

## How it works

The dormant `modifiers.max_per_order` column (declared in SCHEMA and
present in every migrated database, but never read or written) carries
the cap. No migration was needed.

- `triggerQuantity` in `src/shared/db/modifier-resolve.ts` clamps an
  answer trigger's requested quantity to `max_per_order` when set.
  Everything downstream derives from that one choke point: the price
  delta applies at most N times, stock counts orders, the sold-out
  check compares one per order, and `modifier_usages` records the
  capped quantity.
- A stored cap that is not a whole number of 1 or more makes the
  resolve throw (`checkedPerOrderCap`). The column has no CHECK
  constraint, so a repaired or imported row can carry 0. A silent
  undercharge must not follow, so checkout and the webhook oracle both
  refuse the row until an operator repairs it.
- The admin form gains a **Max times per order (optional)** number
  field. Its label and hint live in the message catalog
  (`modifiers.max_per_order.*`) behind `t()`; the i18n allowlist for
  the file stays at its pre-PR count of 20. Save validation requires a
  whole number of 1 or more; `extractModifierInput` keeps the cap only
  for answer triggers — the same rule the promo code field follows — so
  switching a modifier away from that trigger clears a stale cap.
- Public checkout and the webhook price oracle share the same resolve
  path, so both stay consistent automatically.
- The guide recipe points at the field (set it to 1 for a delivery
  fee), and the TODO entry ships with it.

## Tests

- `test/shared/db/modifier-resolve/answers.test.ts` (new) holds the
  answer-trigger suites: quantity clamps to the cap across many tickets
  and across several linked answers (caps 1 and 2 pinned as separate
  tests), stock and oversubscription count per order, the corrupt-cap
  refusals, and the answer link storage/reads. `modifier-resolve.test.ts`
  lands at 358 lines.
- The cap drops for other triggers; 0 and fractional values are
  rejected at save and refused from a stored row. Each new behavior
  test was confirmed to fail with the clamp or validation reverted.
- Mutation: 100% kill across every changed source file
  (`modifier-resolve.ts`, `modifiers.ts`, `features/admin/modifiers.ts`,
  `fields/modifier.ts`, `admin/modifiers/values.ts`, `types.ts`). The
  `types.ts` day-count survivors are recorded equivalent with the
  write-path proof; three of the four became killable when main's
  richer day-prices suite arrived, and their entries are gone.
- Merging main superseded this branch's `types.ts` guard and helper
  suites: main's `guards`/`day-prices`/`shared-groups` files cover every
  claim, richer. The branch's copies are deleted rather than kept
  beside theirs.
- `nix develop -c deno task precommit` and `precommit:mutation` green
  (565/565 killed, 13 known-equivalent suppressed).

## Review

All CodeRabbit and Codex threads are resolved: the withdrawn
length-assertion finding, the catalog migration for the new field
copy, the test-suite split under the 400-line target, the
per-scenario cap tests, and the stored-cap refusal.

## Summary by CodeRabbit

* **New Features**
  * Added an optional **Max times per order** setting for answer-triggered modifiers.
  * Modifiers can now be limited to a specific number of applications per booking, or left unlimited.
  * Delivery-area pricing now supports charging once per booking or per person through this setting.

* **Bug Fixes**
  * Invalid limits, including zero, fractional, or negative values, are now rejected.
  * Limits are ignored for modifier types where they do not apply.

* **Documentation**
  * Updated modifier and delivery-pricing guidance.

### CODE REVIEW

File: test/shared/db/modifier-resolve.test.ts [hunk 1]
```diff
@@ -243,41 +270,128 @@ describeWithEnv("db > modifier-resolve", { db: true }, () => {
...
+      count: number,
+      input: Partial<ModifierInput> = {},
+    ): Promise<{ answerIds: number[]; modifierId: number }> => {
+      const answerIds = await createAnswers(count);
+      const m = await insertModifier(input);
       await patchModifier(m.id, { trigger: "answer" });
-      await setModifierAnswers(m.id, [answerId!]);
+      await setModifierAnswers(m.id, answerIds);
+      return { answerIds, modifierId: m.id };
+    };
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/db/modifier-resolve.test.ts]: Checked, and declined as a perfect-world finding: `createAnswers` builds its return with a `for (let i = 0; i < count; i++) ids.push(...)` loop, so the array length is `count` by construction — an assertion restating that would be tautological (rule: tests must not assert what setup just did). The `answerIds[0]!` / `answerIds[1]!` accesses sit inside tests that immediately use those ids through the production queries, so a missing id would fail the test loudly with an undefined-id error, not silently pass. No concrete input makes the unchecked index wrong here.

File: test/scripts/i18n-coverage.test.ts [hunk 2]
```diff
@@ -86,7 +86,7 @@ const LEFTOVER_ALLOWLIST = new Map<string, number>([
   ["ui/templates/email/defaults.ts", 12],
   ["ui/templates/fields/add-attendee.ts", 3],
   ["ui/templates/fields/listing.ts", 4],
-  ["ui/templates/fields/modifier.ts", 20],
+  ["ui/templates/fields/modifier.ts", 22],
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/scripts/i18n-coverage.test.ts]: Valid. Fixed in f26bd2b63: the label and hint now live in the catalog as `modifiers.max_per_order.label` / `modifiers.max_per_order.hint` and the field reads them through `t()` (the file already resolves picklist labels that way). The allowlist entry for `ui/templates/fields/modifier.ts` ratchets back down to 20, so the file is no deeper in debt than before this PR.

File: test/shared/db/modifier-resolve.test.ts [hunk 3]
```diff
@@ -243,41 +270,132 @@ describeWithEnv("db > modifier-resolve", { db: true }, () => {
       expect(specs.find((s) => s.name === "Limited tee")?.quantity).toBe(2);
     });
 
-    test("applies an answer-triggered modifier when a linked answer is selected", async () => {
-      const [answerId] = await createAnswers(1);
-      const m = await insertModifier({ name: "Large size" });
+    /** Create an answer-triggered modifier linked to `count` fresh answers. */
+    const setUpAnswerModifier = async (
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/db/modifier-resolve.test.ts]: Valid. Fixed in f26bd2b63: all answer-trigger tests and their setup helpers moved to `test/shared/db/modifier-resolve/answers.test.ts` (385 lines) beside the existing `counting.test.ts` / `reachability.test.ts`; `modifier-resolve.test.ts` lands at 358 lines. Along the way the move surfaced duplication of my own making — buyerVisits edge cases and the one-penny add-on test that `counting.test.ts` already covered — so those duplicates are deleted and the two unique buyerVisits cases ("no details at all", "highest count across both details") folded into `counting.test.ts`.

File: test/shared/db/modifier-resolve.test.ts [hunk 4]
```diff
@@ -243,41 +270,132 @@ describeWithEnv("db > modifier-resolve", { db: true }, () => {
...
-    test("caps an answer-triggered modifier quantity at remaining stock", async () => {
-      const [answerId] = await createAnswers(1);
-      const m = await insertModifier({ name: "Limited tier", stock: 2 });
-      await patchModifier(m.id, { trigger: "answer" });
-      await setModifierAnswers(m.id, [answerId!]);
-
-      const specs = await resolveModifiers([checkoutItem()], {
-        answerQuantities: await answerModifierQuantities(
-          { "1": [answerId!] },
+    test("caps an answer modifier at stock, or once per order when capped", async () => {
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/db/modifier-resolve.test.ts]: Valid. Fixed in f26bd2b63: the three scenarios are now separate tests registered through a small `capsQuantityAt(description, input, expected)` factory (the same registration shape as `checkBothArms`), so each has its own name and a single failure reason, while the shared body lives once — which is also what keeps jscpd at 0%, the reason the cases were originally merged.

**Comment by stefan-burke** [comment]
@coderabbitai review

File: src/shared/db/modifier-resolve.ts [hunk 5]
```diff
@@ -144,7 +146,10 @@ const triggerQuantity = (
   }
   if (modifier.trigger === "optional") return addOns.get(modifier.id) ?? 0;
   if (modifier.trigger === "answer") {
-    return answerQuantities.get(modifier.id) ?? 0;
+    const requested = answerQuantities.get(modifier.id) ?? 0;
+    return modifier.max_per_order === null
+      ? requested
+      : Math.min(requested, modifier.max_per_order);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/modifier-resolve.ts]: Valid, and it matched the repo rule that an observed impossible state must be raised, not normalised — a cap of 0 turned a corrupt row into a silent undercharge. Fixed in cf033b676: `triggerQuantity` now routes every non-null stored cap through `checkedPerOrderCap`, which throws when the value is not a whole number of 1 or more (checkout and the webhook price oracle share this resolve path, so both refuse it). Regression tests `refuses a stored per-order cap of zero` and `refuses a fractional stored per-order cap` in `test/shared/db/modifier-resolve/answers.test.ts` write the bad row the way a repair or import would (`patchModifier` straight to the column) and assert the resolve rejects — both fail on the pre-fix code.

File: src/shared/db/modifier-resolve.ts [hunk 6]
```diff
@@ -128,11 +128,26 @@ const stockedQuantity = (
...
 
+/** A stored `max_per_order` the resolve path may apply. The column has no
+ * CHECK constraint, so a repaired or imported row can carry a cap the admin
+ * form refuses. A cap of 0 would silently drop the charge from every order,
+ * so refuse it loudly instead. */
+const checkedPerOrderCap = (modifier: Modifier, cap: number): number => {
+  if (!Number.isInteger(cap) || cap < 1) {
+    throw new Error(
+      `Modifier "${modifier.name}" (id ${modifier.id}) stores max_per_order ${cap}; the column must hold a whole number of 1 or more`,
+    );
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/modifier-resolve.ts:141]: Valid — Simplified Technical English forbids the semicolon in developer-facing text. Fixed in a8aa63d18: the refusal is two sentences now (`... stores max_per_order 0. The column must hold a whole number of 1 or more`). The regression tests assert on the `max_per_order` substring, so they still pin the refusal.

File: src/locales/en/guide-operations.json [hunk 7]
```diff
@@ -62,7 +62,7 @@
...
+  "guide.a.charge_delivery_by_area": "<p>Set one price for each area you deliver to. The buyer picks their area on the booking form. The area's price is added to what they pay.</p><ol><li>Create a question such as &quot;Which delivery area?&quot;, with one answer per area &mdash; for example &quot;Zone 1&quot; and &quot;Zone 2&quot;. Assign the question to the listings you deliver.</li><li>Open <a href=\"/admin/modifiers\">Modifiers</a> and add a modifier for each area. Set the trigger to <strong>Question answer</strong>, tick the answer for that area, and give it a fixed charge with the area's price.</li><li>Set <strong>Max times per order</strong> to 1 on the modifier, so the fee is charged once per booking instead of for every ticket.</li></ol><p>The price comes from the area the buyer picks, not from their postcode &mdash; like a price list, not a distance calculation. Leave <strong>Max times per order</strong> blank and the price works like a per-person delivery charge. Set a modifier's scope when the charge must apply to only some listings.</p>",
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/locales/en/guide-operations.json]: Valid — one idea per sentence. Fixed in a8aa63d18: the step reads "Set **Max times per order** to 1 on the modifier. The fee is charged once per booking, not for every ticket." The neighbouring sentence in the same entry joined two ideas with "and", so it splits the same way: "Leave **Max times per order** blank. The price then works like a per-person delivery charge."

File: src/features/admin/modifiers.ts [hunk 8]
```diff
@@ -219,6 +222,15 @@ const validateModifier = (
...
     );
   }
+  if (
+    input.maxPerOrder !== undefined &&
+    input.maxPerOrder !== null &&
+    (!Number.isInteger(input.maxPerOrder) || input.maxPerOrder < 1)
+  ) {
+    return Promise.resolve(
+      "Max times per order must be a whole number of 1 or more",
+    );
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/admin/modifiers.ts:232]: Declined as tracked work rather than a fix for this PR — the routes-to-catalog migration owns this file whole. `TODO.md` on main names `modifiers.ts` in "Move the routes' flash messages into the message catalog": twenty files under `src/features/`, 41 literals, and the job includes widening the i18n-coverage check to `src/features/` so a new one cannot appear. Every sibling message in this validator is a literal of the same shape ("Minimum previous bookings must be a whole number of 0 or more" two lines up), and the new message copies that shape on purpose, so the file reads one way. Migrating only the new key now would split the file between two mechanisms and take one bite out of a job that belongs whole — the catalog rule for this file lands with that entry, checker included.

File: src/shared/db/modifier-resolve.ts [hunk 9]
```diff
@@ -128,11 +128,26 @@ const stockedQuantity = (
...
+ * form refuses. A cap of 0 would silently drop the charge from every order,
+ * so refuse it loudly instead. */
+const checkedPerOrderCap = (modifier: Modifier, cap: number): number => {
+  if (!Number.isInteger(cap) || cap < 1) {
+    throw new Error(
+      `Modifier "${modifier.name}" (id ${modifier.id}) stores max_per_order ${cap}. The column must hold a whole number of 1 or more`,
+    );
+  }
+  return cap;
+};
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/modifier-resolve.ts]: Valid — one corrupt row was able to reject pricing for every cart, and that is a concrete availability failure under the same repair/import scenario the original finding named. Fixed in 985d8bfb0: the cap is read only when the order picked a linked answer (`if (requested < 1 || cap === null) return requested;`), so unrelated carts price normally while the orders the cap would mischarge still refuse loudly. Two tests pin it: "refuses a corrupt cap when one ticket picked the answer" holds the boundary at the smallest order that pays, and "does not read the cap of an answer modifier no order picked" proves an unselected corrupt row prices normally. Mutation gate re-run green (565/565).

File: test/shared/db/modifier-resolve.test.ts [hunk 10]
```diff
@@ -93,6 +71,27 @@ describe("childUnreachableAddOnError", () => {
...
+  });
+
+  test("reachable pages are the active listings that are not children", () => {
+    expect(
+      reachablePageIds(
+        [{ active: true, id: 1 }, { id: 2 }, { active: true, id: 3 }],
+        new Set([3]),
+      ),
+    ).toEqual(new Set([1]));
+  });
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/db/modifier-resolve.test.ts:96]: Valid — the failure would report under the wrong function name. Moved to its own `describe("reachablePageIds")` block in 985d8bfb0.

File: test/shared/db/modifier-resolve/answers.test.ts [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+    });
+
+    test("answerModifierQuantities ignores answers with no linked modifier", async () => {
+      const [answerId] = await createAnswers(1);
+      expect(
+        await answerModifierQuantities({ "1": [answerId!] }, new Map([[1, 4]])),
+      ).toEqual(new Map());
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/shared/db/modifier-resolve/answers.test.ts:209]: Valid — my corrupt-cap tests pushed the file to 435 lines. Split: the link-storage, `answerModifierQuantities`, and `oversubscribedAnswerTiers` suites moved to `answer-links.test.ts` (227 lines), the resolve/cap suites stay in `answers.test.ts` (209), and the shared setup (`createAnswers`, `setUpAnswerModifier`, `resolveAnswerPicks`) lives once in `answer-setup.ts` beside them, so jscpd stays at 0%.

File: src/features/admin/modifiers.ts [hunk 12]
```diff
@@ -86,7 +86,9 @@ import { makeMoneyAdjustHandler } from "./money-adjust.ts";
 /** Build modifier input from validated form values. The value is stored as the
  * positive magnitude the owner typed; converting it to the signed engine value
  * happens where modifiers are applied to a checkout. A promo code is kept only
- * for "code" modifiers, with its blind index computed for public lookup. */
+ * for "code" modifiers, with its blind index computed for public lookup; the
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/admin/modifiers.ts]: Valid — both semicolons are gone. The `types.ts` cap JSDoc is three sentences now, and the `extractModifierInput` JSDoc also drops its gerund openers ("converting it… happens", "switching a modifier… clears") for plain sentences, which the same STE rule forbids.

File: src/shared/db/modifier-resolve.ts [hunk 13]
```diff
@@ -144,7 +161,10 @@ const triggerQuantity = (
   }
   if (modifier.trigger === "optional") return addOns.get(modifier.id) ?? 0;
   if (modifier.trigger === "answer") {
-    return answerQuantities.get(modifier.id) ?? 0;
+    const requested = answerQuantities.get(modifier.id) ?? 0;
+    const cap = modifier.max_per_order;
+    if (requested < 1 || cap === null) return requested;
+    return Math.min(requested, checkedPerOrderCap(modifier, cap));
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/modifier-resolve.ts:167]: Valid — same blast-radius fault one gate further in. `eligibleCandidates` now runs the visit, scope, and subtotal gates before it computes the quantity, so a corrupt cap refuses only carts the modifier would really apply to. New test "does not read the cap of a modifier the cart is too small for" pins it (corrupt cap behind an unmet `min_subtotal` prices normally). Mutation gate re-run green.
